### ISSUE TITLE
DL4J: LSTM MKLDNN Support

### ISSUE THREAD

**Issue by AlexDBlack** [eclipse/deeplearning4j#8051]
We have it for conv, subsampling, etc but no LSTM...

**Comment by AlexDBlack** [comment]
OK, so looking at MKLDNN LSTMs:

https://intel.github.io/mkl-dnn/group__c__api__rnn.html#gaadc3358de256c75ee19db6d2707869fe
https://github.com/intel/mkl-dnn/blob/master/examples/cpu_rnn_training_f32.cpp

https://intel.github.io/mkl-dnn/dev_guide_rnn.html
Note also the data format information at the end.


Weights must be together in one one array for input, a sepate one for recurrent, matching DL4J.
Order is "IFOG" aka "afco" in MKLDNN docs (input, forget, input modulation, output), which matches DL4J RNNs.
Unlike DL4J, it's hardcoded tanh/sigmoid, and no peephole connections.

Unfortunately, it does not appear that we can use the lstmBlock op here (it's necessary to retain this format for TF import):
https://github.com/eclipse/deeplearning4j/blob/master/libnd4j/include/ops/declarable/generic/recurrent/lstmBlock.cpp
Differences in weights (lstmBlock combines input/recurrent weights together into one).
More significantly, the outputs are quite different: lstmBlock has 7 outputs (full activations for all steps);
MKLDNN seems to do a more sensible thing and only output the final activations, and the final steps for (I assume) the last time step activations and last cell state activations.

From what I recall from cuDNN, this is the same as cuDNN.


So:
1. We should leave lstmBlock op as-is, and not use it here. Keep it just for import only.
2. We need a new LSTM op (and backprop op) in libnd4j, that can be used for MKLDNN and (later cuDNN) that we'll use within DL4J.

**Comment by saudet** [comment]
BTW, the API changed somewhat with MKL-DNN 1.0. Let's make sure we don't implement something for the old API.

**Comment by AlexDBlack** [comment]
@saudet thanks, I'll double check that

Also: MKLDNN looks to essentially match ONNX as well (though ONNX supports peephole connections), so the same op should be the "main" one with (depending on args) MKLDNN/cuDNN/ONNX support.
However we'll need a non-MKLDNN/non-cuDNN version for peepholes and activation functions that 

https://github.com/onnx/onnx/blob/master/docs/Operators.md#lstm

**Comment by AlexDBlack** [comment]
MKLDNN transition guide - https://intel.github.io/mkl-dnn/dev_guide_transition_to_v1.html

A few changes of note impacting RNNs, nothing too major though.

Perhaps it would make sense to upgrade to MKL-DNN 1.0 first? (If so, I'll open an issue for that)
@saudet I didn't see any javacpp presets for MLK-DNN 1.0... is that already planned/on the roadmap?

**Comment by saudet** [comment]
It's in the `mkl-dnn` branch: https://github.com/bytedeco/javacpp-presets/tree/mkl-dnn
It doesn't come with MKLML anymore, so until other projects also upgrade (MXNet, TensorFlow, and nGraph), we'll need to deploy it separately, but that's not a big issue.

**Comment by raver119** [comment]
LSTM op added, MKL-DNN updated to 1.0.x, LSTM bp is wip.
